exporters/prometheus/src/collector.cc

```diff
@@ -36,8 +39,19 @@ std::vector<prometheus_client::MetricFamily> PrometheusCollector::Collect() cons
   collection_lock_.lock();
 
   std::vector<prometheus_client::MetricFamily> result;
-  reader_->Collect([&result](sdk::metrics::ResourceMetrics &metric_data) {
-    auto prometheus_metric_data = PrometheusExporterUtils::TranslateToPrometheus(metric_data);
+
+  // If populate_target_info_ is true, populate target_info once.
+  bool populate_target_info = false;
+  if (populate_target_info_)
+  {
+    populate_target_info = true;
+
+    const_cast<PrometheusCollector *>(this)->populate_target_info_ = false;
```

Comment:
> There MUST be at most one target_info metric exposed on an SDK Prometheus endpoint.
According to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#resource-attributes-1 , the target info metric should not change and it must be exposed only once.Do I missed anything?

Reply:
This means once per-scrape/collection.  I.e. there should never be a target_info with two streams:

```
# TYPE target info
# HELP target Target metadata
target_info{env="prod",hostname="myhost",datacenter="sdc",region="europe",owner="frontend"} 1
target_info{env="staging",hostname="myhost",datacenter="sdc",region="europe",owner="frontend"} 1
```

You should have the target info metric on every scrape.

Reply:
I guess this was modified for the comment [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/2301#discussion_r1329652826).  Probably I was not very clear - the ask was to compute and cache the `target_info` metric internally during first scrape(as resources are immutable), and add it during every scrape/collection.

Reply:
+1.  Caching is great.

Reply:
The referenced line disables sending target_info only if we encounter an error when creating the target_info metric.  If we failed to create the target_info metric, it will be nil, so we shouldn't try and send it. In the normal, non-error case, otel-go caches the target_info metric (as c.target_info), and [adds it on every scrape](https://github.com/open-telemetry/opentelemetry-go/blob/9bbefc6cc31995c7bdd37c69eb153fc2201c3562/exporters/prometheus/exporter.go#L178).

You are welcome to follow the same pattern of disabling target_info if creating the metric fails, but please do not only add target_info on the first scrape.

Reply:
@owent - It seems that sending target_info with every scrape should be corrected, right? I'd like to revisit the PR once this is addressed, so checking.

Reply:
> @owent - It seems that sending target_info with every scrape should be corrected, right? I'd like to revisit the PR once this is addressed, so checking.

Fixed, sorry for forgot it.